client/components/domains/contact-details-form-fields/index.jsx

```diff
@@ -144,6 +144,8 @@ export class ContactDetailsFormFields extends Component {
 		const mainFieldValues = formState.getAllFieldValues( this.state.form );
 		return {
 			...mainFieldValues,
+			// domains registered according to ancient validation rules may have state set even though not required
+			state: this.props.hasCountryStates ? mainFieldValues.state : '',
```

Comment:
Are you sure `this.props.hasCountryStates` is the correct check for this? AFAICT, it checks if we have a strict list of allowed states for the given country, not that the state is set at all (which I think is what should be the checked here, unless I'm mistaken).

Reply:
> AFAICT, it checks if we have a strict list of allowed states for the given country

That's how I understand it too.

> not that the state is set at all (which I think is what should be the checked here, unless I'm mistaken)

I was just clearing it if not required. Do we care if the address has a previously-filled state/province for countries that no longer require it during registration? 

If so, for WHOIS form users in UK/IE/EU we could check for a state value, and show the US-style version of the form if there is one. The only catch is that if they go ahead and register a new domain, the state field won't be there which might be confusing. Or it might not be. :)

Furthermore, I see there's state name validation in the backend to handle special characters, but it allows `a-z A-Z 0-9 . , \' - [space]` which would permit another `p2MSmN-6p4-p2` situation, no?

Reply:
Ah, I misunderstood your intention here then (and how the code works). If we have a strict state list then that implies that the state is required, right? And if we don't have a strict state list, we _never_ require the state field? Is that correct? The last part is what I'm confused about - I'm not sure if it's true, but you have more experience with it, so I'll defer to you.
Other than that, I don't have an issue with this - the less fields we need to show to the user, the better.

Reply:
> If we have a strict state list then that implies that the state is required, right?

Good question. From what I tell, we have a state name validator whose first check is **_"if the state is empty, and the user's country code has a corresponding state list, then flag the state field as required"_**.

@deBhal could you confirm this?

Reply:
Huh, TIL. I thought those states were primarily a UI convenience to populate dropdowns, but I found r117676-wpcom which says:

> "WWD expect an English state name for specific countries whereas it is not required for other countries."

So that reads to me that this list is both a whitelist and a list of requirements.

> if we don't have a strict state list, we never require the state field?

That seems right, yes. We've haven't had to implement any other state validations because they're all looser than this one. `.CA ` for example requires a state for the US and Canada, but we didn't write anything because it was already covered by this check.

> the less fields we need to show to the user, the better.
💯